tools/sslsniff.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python
+#
+# sslsniff  Captures data on SSL_READ or SSL_WRITE functions of OpenSSL
+#           For Linux, uses BCC, eBPF.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 12-Aug-2016    Adrian Lopez   Created this.
+# 13-Aug-2016    Mark Drayton   Fix SSL_Read
+
+import ctypes as ct
+from bcc import BPF
+
+prog = """
+#include <linux/ptrace.h>
+#include <linux/sched.h>        /* For TASK_COMM_LEN */
+
+struct probe_SSL_data_t
+{
+        u64 timestamp_ns;
+        u32 pid;
+        char comm[TASK_COMM_LEN];
+        char v0[476];
+};
+
+BPF_PERF_OUTPUT(perf_SSL_write);
+
+int probe_SSL_write(struct pt_regs *ctx)
+{
+        struct probe_SSL_data_t __data = {0};
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = bpf_get_current_pid_tgid();
+
+        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
+
+        int ret;
+
+        if ( (void *) PT_REGS_PARM2(ctx) != 0) {
+                bpf_probe_read(&__data.v0, sizeof(__data.v0), (void *) ctx->si);
+                ret = bpf_probe_read(&__data.v0, sizeof(__data.v0), (unsigned char*) PT_REGS_PARM2(ctx));
+        }
+
+        perf_SSL_write.perf_submit(ctx, &__data, sizeof(__data));
+        return 0;
+}
+
+BPF_PERF_OUTPUT(perf_SSL_read);
+
+BPF_HASH(bufs, u32, u64);
+
+int probe_SSL_read_enter(struct pt_regs *ctx)
+{
+        u32 pid = bpf_get_current_pid_tgid();
+        u64 buf = (u64)PT_REGS_PARM2(ctx);
+        bufs.update(&pid, &buf);
+        return 0;
+}
+
+int probe_SSL_read_exit(struct pt_regs *ctx) {
+        u32 pid = bpf_get_current_pid_tgid();
+        u64 *bufp = bufs.lookup(&pid);
+        if (bufp == 0) {
+                return 0;
+        }
+
+        struct probe_SSL_data_t __data = {0};
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = pid;
+
+        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
+
+        int ret;
+        if (bufp != 0) {
+                ret = bpf_probe_read(&__data.v0, sizeof(__data.v0), (char *)*bufp);
+        }
+
+        bufs.delete(&pid);
+
+        perf_SSL_read.perf_submit(ctx, &__data, sizeof(__data));
+        return 0;
+}
+"""
+
+b = BPF(text=prog)
+libpath = "/usr/lib/libssl.so.1.0.0"
+
+# Join to ssl_write
+b.attach_uprobe(name=libpath, sym="SSL_write", fn_name="probe_SSL_write")
```

Comment:
You can just use `name="ssl"` here. It'll be resolved into the correct path to libssl (which may differ from yours) automagically.